File: src/php/Optimization/ProblemInitializers.php
<?php

declare(strict_types=1);

namespace Anonympins\Fingerprint\Optimization;

/**
 * Generators for dynamically seeding problem instance datasets.
 */
class ProblemInitializers
{
    /** @var array<string, callable> */
    private static array $initializers = [];

    private static function randomFloat(): float
    {
        try {
            return (float)random_int(0, PHP_INT_MAX - 1) / (float)PHP_INT_MAX;
        } catch (\Throwable $e) {
            if (function_exists('wp_rand')) {
                return (float)wp_rand(0, 1000000) / 1000000.0;
            }
            return 0.5;
        }
    }

    private static function initialize(): void
    {
        if (empty(self::$initializers)) {
            self::$initializers['generate:randomPoints'] = function (array $params): array {
                $count = $params['count'] ?? 0;
                $bounds = $params['bounds'] ?? ['x' => 1000, 'y' => 1000];
                if (!is_numeric($count)) return [];
                $points = [];
                for ($i = 0; $i < $count; $i++) {
                    $points[] = [
                        'x' => self::randomFloat() * $bounds['x'],
                        'y' => self::randomFloat() * $bounds['y']
                    ];
                }
                return $points;
            };

            self::$initializers['generate:randomAssets'] = function (array $params): array {
                $count = $params['count'] ?? 0;
                if (!is_numeric($count)) return [];
                $assets = [];
                for ($i = 0; $i < $count; $i++) {
                    $assets[] = [
                        'name' => 'Asset ' . ($i + 1),
                        'expectedReturn' => self::randomFloat() * 0.2,
                        'volatility' => 0.1 + self::randomFloat() * 0.3
                    ];
                }
                return $assets;
            };
        }
    }

    public static function get(string $name): ?callable
    {
        self::initialize();
        return self::$initializers[$name] ?? null;
    }
}

File: src/java/src/com/anonympins/fingerprint/utils/ProblemInitializers.java
package com.anonympins.fingerprint.utils;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Random;
import java.util.function.Function;

/**
 * Fonctions pour générer dynamiquement les données d'un problème.
 */
public class ProblemInitializers {
    private static final Map<String, Function<Map<String, Object>, Object>> initializers = new HashMap<>();

    static {
        // Example: generate:randomPoints for TSP
        initializers.put("generate:randomPoints", params -> {
            int count = ((Number) params.getOrDefault("count", 0)).intValue();
            Map<String, Object> bounds = (Map<String, Object>) params.getOrDefault("bounds", new HashMap<>());
            double maxX = ((Number) bounds.getOrDefault("x", 1000.0)).doubleValue();
            double maxY = ((Number) bounds.getOrDefault("y", 1000.0)).doubleValue();
            
            List<Map<String, Double>> points = new ArrayList<>();
            Random rand = new Random();
            for (int i = 0; i < count; i++) {
                Map<String, Double> point = new HashMap<>();
                point.put("x", rand.nextDouble() * maxX);
                point.put("y", rand.nextDouble() * maxY);
                points.add(point);
            }
            return points;
        });

        // Example: generate:randomAssets for Portfolio Optimization
        initializers.put("generate:randomAssets", params -> {
            int count = ((Number) params.getOrDefault("count", 0)).intValue();
            List<Map<String, Object>> assets = new ArrayList<>();
            Random rand = new Random();
            for (int i = 0; i < count; i++) {
                Map<String, Object> asset = new HashMap<>();
                asset.put("expectedReturn", rand.nextDouble() * 0.2);
                asset.put("volatility", 0.1 + rand.nextDouble() * 0.3);
                assets.add(asset);
            }
            return assets;
        });
    }

    public static Function<Map<String, Object>, Object> getInitializer(String name) {
        return initializers.get(name);
    }
}